src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutVertexNode.cs

```diff
@@ -0,0 +1,405 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.NativeFormat;
+using Internal.TypeSystem;
+using ILCompiler.DependencyAnalysisFramework;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Wrapper nodes for native layout vertex structures. These wrapper nodes are "abstract" as they do not 
+    /// generate any data. They are used to keep track of the dependency nodes required by a Vertex structure.
+    /// 
+    /// Any node in the graph that references data in the native layout blob needs to create one of these
+    /// NativeLayoutVertexNode nodes, and track it as a dependency of itself.
+    /// Example: MethodCodeNodes that are saved to the table in the ExactMethodInstantiationsNode reference 
+    /// signatures stored in the native layout blob, so a NativeLayoutPlacedSignatureVertexNode node is created
+    /// and returned as a static dependency of the associated MethodCodeNode (in the GetStaticDependencies API).
+    /// 
+    /// Each NativeLayoutVertexNode that gets marked in the graph will register itself with the NativeLayoutInfoNode,
+    /// so that the NativeLayoutInfoNode can write it later to the native layout blob during the call to its GetData API.
+    /// </summary>
+    internal abstract class NativeLayoutVertexNode : DependencyNodeCore<NodeFactory>
+    {
+        private Vertex _savedVertex;
+
+        public Vertex SavedVertex
+        {
+            get { return _savedVertex; }
+            protected set
+            {
+                Debug.Assert(_savedVertex == null || Object.ReferenceEquals(_savedVertex, value));
+                _savedVertex = value;
+            }
+        }
+
+        public override bool HasConditionalStaticDependencies => false;
+        public override bool HasDynamicDependencies => false;
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+        public override bool StaticDependenciesAreComputed => true;
+        protected override string GetName() => null;
+
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory context)
+        {
+            return Array.Empty<CombinedDependencyListEntry>();
+        }
+
+        public override IEnumerable<CombinedDependencyListEntry> SearchDynamicDependencies(List<DependencyNodeCore<NodeFactory>> markedNodes, int firstNode, NodeFactory context)
+        {
+            return Array.Empty<CombinedDependencyListEntry>();
+        }
+
+        protected override void OnMarked(NodeFactory context)
+        {
+            context.MetadataManager.NativeLayoutInfo.AddVertexNodeToNativeLayout(this);
+        }
+
+        public abstract Vertex WriteVertex(NodeFactory factory, NativeWriter writer);
+    }
+
+    internal sealed class NativeLayoutMethodLdTokenVertexNode : NativeLayoutVertexNode
+    {
+        private MethodDesc _method;
+        private NativeLayoutTypeSignatureVertexNode _containingTypeSig;
```

Comment:
We don't generally cache things that we can easily get from the `NodeFactory`.

`NodeFactory` is a parameter to most of the interesting methods on nodes because we don't want nodes to carry too much redundant state.